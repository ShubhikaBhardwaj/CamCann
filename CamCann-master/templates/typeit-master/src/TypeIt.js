import Instance from "./Instance";
import allHaveStatus from "./helpers/allHaveStatus";
import { generateHash } from "./utilities";
import toArrayOfNodes from "./helpers/toArrayOfNodes";

export default class TypeIt {
  constructor(element, options) {
    this.instances = toArrayOfNodes(element).map(el => {
      return new Instance({
        typeIt: this,
        element: el,
        id: generateHash(),
        options,
        queue: []
      });
    });
  }

  each(func) {
    this.instances.forEach(instance => {
      func.call(this, instance);
    });
  }

  /**
   * Push a specific action directly into the queue of each instance.
   * If an instance has already completed, trigger the queeu again.
   *
   * @param {string} function
   * @param {*} argument
   * @
   */
  queueUp(action, argument = null, numberOfTimesToCopy = 1) {
    this.each(instance => {
      let isIndependentFunction = typeof action !== "string";

      /**
       * If action is a string, bind it to instance.
       * Otherwise, leave it on its own.
       */
      let toFire = isIndependentFunction ? action : instance[action];

      let toPassAsArguments = isIndependentFunction ? this : argument;

      for (let i = 0; i < numberOfTimesToCopy; i++) {
        instance.queue.add([toFire, toPassAsArguments]);
      }
    });
  }

  is(status) {
    return allHaveStatus(this.instances, status, true);
  }

  freeze() {
    this.each(instance => {
      instance.status.frozen = true;
    });
  }

  unfreeze() {
    this.each(instance => {
      if (!instance.status.frozen) return;
      instance.status.frozen = false;
      instance.fire();
    });
  }

  /**
   * If used after typing has started, will append strings to the end of the existing queue. If used when typing is paused, will restart it.
   *
   * @param  {string} string The string to be typed.
   * @return {object} TypeIt instance
   */
  type(string = "") {
    this.each(instance => instance.queueString(string));
    return this;
  }

  /**
   * If null is passed, will delete whatever's currently in the element.
   *
   * @param  { number } numCharacters Number of characters to delete.
   * @return { TypeIt }
   */
  delete(numberOfCharactersToDelete = null) {
    this.queueUp(
      "delete",
      numberOfCharactersToDelete === null, //-- Maybe delete all.
      numberOfCharactersToDelete === null ? 1 : numberOfCharactersToDelete
    );

    return this;
  }

  pause(ms = null) {
    this.queueUp("pause", ms);
    return this;
  }

  break() {
    this.queueUp("type", document.createElement("BR"));
    return this;
  }

  options(options) {
    this.queueUp("setOptions", options);
    return this;
  }

  exec(func) {
    this.queueUp(func);
    return this;
  }

  /**
   * Destroy the instance, mark as such, and clean up. If specified,
   * remove the cursor on destruction.
   *
   * @param {boolean} removeCursor
   */
  destroy(removeCursor = true) {
    this.each(instance => {
      // Destroy each timeout, for good housekeeping.
      instance.timeouts.forEach(timeout => {
        clearTimeout(timeout);
      });

      instance.timeouts = [];

      // Clean up cursor node, if specified.
      if (removeCursor) {
        let cursorNode = instance.isInput
          ? null
          : instance.$e.querySelector(".ti-cursor");

        if (cursorNode) {
          instance.$e.removeChild(cursorNode);
        }
      }

      instance.status.destroyed = true;
    });
  }

  empty() {
    this.queueUp("empty");
    return this;
  }

  /**
   * Reset each instance like it's brand new.
   */
  reset() {
    this.destroy();

    this.instances = this.instances.map(instance => {
      return instance.reset();
    });

    return this;
  }

  /**
   * Initialize each instance.
   */
  go() {
    this.each(instance => {
      instance.init();
    });

    return this;
  }
}
