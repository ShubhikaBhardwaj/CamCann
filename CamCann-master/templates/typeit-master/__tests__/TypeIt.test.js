import TypeIt from "../src/TypeIt.js";

let instance;

beforeEach(() => {
  document.body.innerHTML = `
    <div>
      <span id="element"></span>
    </div>
  `;

  instance = new TypeIt("#element");
});

test("Initial queue only contains startDelay pause.", () => {
  expect(instance.instances[0].queue).toMatchSnapshot();
});

test("Returns an object with expected properties.", () => {
  expect(Object.keys(instance).sort()).toEqual(["instances"]);
});

test("Defines hard-coded string correctly.", () => {
  document.body.innerHTML = `
    <div>
      <span id="element">Hard-coded string.</span>
    </div>
  `;

  let instance = new TypeIt("#element", {
    strings: ["My string."]
  });

  expect(instance.instances[0].opts.strings).toEqual([
    "Hard-coded string.",
    "My string."
  ]);
});

test("Will not begin until explicitly called.", () => {
  document.body.innerHTML = `<div>'
      <span id="element"></span>
    </div>`;

  const instance = new TypeIt("#element", {
    strings: "hello!"
  });

  expect(instance.is("started")).toBe(false);

  instance.go();

  expect(instance.is("started")).toBe(true);
});

test("Clears out remnants of previous instances correctly.", () => {
  document.body.innerHTML = `
    <div>
      <span id="element">
        Previous string.
      </span>
    </div>
  `;

  let instance = new TypeIt("#element", {
    strings: "My string."
  });

  expect(!instance.instances[0].opts.strings[0].includes("ti-cursor")).toEqual(
    true
  );
});

test("Typing doesn't end with a break tag.", () => {
  document.body.innerHTML = `<div>'
    <span id="element"></span>
  </div>`;

  const instance = new TypeIt("#element", {
    strings: ["One string.", "Two string", "Three string."]
  }).go();

  expect(instance.instances[0].$e.innerHTML.endsWith("<br>")).not.toBe(true);
});

describe("reset()", () => {
  test("Successfully resets when called.", () => {
    document.body.innerHTML = `<div>'
        <span id="element"></span>
      </div>`;

    const instance = new TypeIt("#element", {
      strings: "This is my string!"
    }).go();

    instance.destroy();

    expect(instance.is("destroyed")).toBe(true);

    instance.reset();

    //-- Ensure the arguments that define these properties were passed.
    expect(instance.instances[0].$e).not.toBe(undefined);
    expect(instance.instances[0].id).not.toBe(undefined);
    expect(instance.instances[0].opts).not.toBe(undefined);
    expect(instance.instances).toHaveLength(1);
    expect(instance.is("completed")).toBe(false);
    expect(instance.is("destroyed")).toBe(false);
  });
});

describe("destroy()", () => {
  test("Destroys instances successfully.", () => {
    document.body.innerHTML = `<div>'
      <span id="element"></span>
    </div>`;

    const instance = new TypeIt("#element", {
      strings: "This is my string!"
    }).go();

    expect(instance.is("destroyed")).toBe(false);

    instance.destroy();

    expect(instance.instances[0].timeouts).toEqual([]);
    expect(document.body.querySelector(".ti-cursor")).toEqual(null);
  });
});

describe("queueUp()", () => {
  test("It should add function name by string w/ correct arguments.", () => {
    instance.queueUp("type", "someArg");
    expect(instance.instances[0].queue.waiting).toMatchSnapshot();
  });

  test("If invalid function name is passed, will save as undefined", () => {
    instance.queueUp("thisDoesNotExist", "someArg");
    expect(instance.instances[0].queue.waiting).toMatchSnapshot();
  });

  test("Correctly enqueues anonymous function.", () => {
    instance.queueUp(function() {
      return true;
    });

    let funcAsString = instance.instances[0].queue.waiting[1][0].toString();
    let argument = instance.instances[0].queue.waiting[1][1];

    expect(funcAsString).toMatchSnapshot();
    expect(argument).toBeInstanceOf(TypeIt);
  });

  test("It duplicates the action a certain number of times per instance, if specified.", () => {
    //-- Add 'type' to queue three times.
    instance.queueUp("type", 1, 3);

    expect(instance.instances[0].queue.waiting).toMatchSnapshot();

    instance.queueUp("delete", 1);

    expect(instance.instances[0].queue.waiting).toMatchSnapshot();
  });
});

describe("each()", () => {
  test("It should fire a function on an instance.", () => {
    let mockCallback = jest.fn(instance => instance.$e);

    instance.each(mockCallback);

    expect(mockCallback.mock.calls[0][0].constructor.name).toEqual("Instance");

    expect(mockCallback.mock.calls.length).toBe(1);
  });

  test("It should fire a function on several instances.", () => {
    document.body.innerHTML = `
      <div>
        <span class="element"></span>
        <span class="element"></span>
        <span class="element"></span>
      </div>
    `;

    instance = new TypeIt(".element");

    let mockCallback = jest.fn(instance => instance.$e);

    instance.each(mockCallback);

    expect(mockCallback.mock.calls[0][0].constructor.name).toEqual("Instance");

    expect(mockCallback.mock.calls.length).toBe(3);
  });
});
