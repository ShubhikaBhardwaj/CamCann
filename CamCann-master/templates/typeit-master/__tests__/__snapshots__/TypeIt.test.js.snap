// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Initial queue only contains startDelay pause. 1`] = `
_default {
  "executed": Array [],
  "waiting": Array [
    Array [
      [Function],
      250,
    ],
  ],
}
`;

exports[`queueUp() Correctly enqueues anonymous function. 1`] = `
"function () {
      return true;
    }"
`;

exports[`queueUp() If invalid function name is passed, will save as undefined 1`] = `
Array [
  Array [
    [Function],
    250,
  ],
  Array [
    undefined,
    "someArg",
  ],
]
`;

exports[`queueUp() It duplicates the action a certain number of times per instance, if specified. 1`] = `
Array [
  Array [
    [Function],
    250,
  ],
  Array [
    [Function],
    1,
  ],
  Array [
    [Function],
    1,
  ],
  Array [
    [Function],
    1,
  ],
]
`;

exports[`queueUp() It duplicates the action a certain number of times per instance, if specified. 2`] = `
Array [
  Array [
    [Function],
    250,
  ],
  Array [
    [Function],
    1,
  ],
  Array [
    [Function],
    1,
  ],
  Array [
    [Function],
    1,
  ],
  Array [
    [Function],
    1,
  ],
]
`;

exports[`queueUp() It should add function name by string w/ correct arguments. 1`] = `
Array [
  Array [
    [Function],
    250,
  ],
  Array [
    [Function],
    "someArg",
  ],
]
`;
